testng-core/src/main/java/org/testng/internal/ClassImpl.java

```diff
@@ -94,7 +95,13 @@ private Object getDefaultInstance(boolean create, String errMsgPrefix) {
       if (m_instance != null) {
         m_defaultInstance = m_instance;
       } else {
-        IObjectDispenser dispenser = Dispenser.newInstance(m_objectFactory);
+        ITestObjectFactory factory = m_objectFactory;
+        Class<?> factoryClass = factory.getClass();
+        if (factoryClass.isAnonymousClass()
```

Comment:
I think it will be easier to use a real class

Reply:
No. Sometimes we get an anonymous object from within TestNG and thats why I have put this logic to ensure that only if I stumble into an anonymous object and whose enclosing class is TestNG, we resort to this logic.

Reply:
That's the point. Replace anonymous objects with real classes when needed. I don't think we will remember this side effect if someone will do it later for any reason.
Using a real class enforces the unbreakable logic.

Reply:
Fair enough. I will refactor that part also to use a proper class and then have this changed accordingly.

Reply:
This is now fixed.